### PR TITLE
Convert calculated pool quantities to columns

### PR DESCRIPTION
**Pull request by wottop**
Adds columns and migrates data on db update.
Programmatically increments and decrements pool
quantities based on entitling.

### CODE REVIEW

**Comment by barnabycourt** [comment]
@wottop The pool quantity will need to be modified on entitlement deletion as well as creation.

**Comment by wottop** [comment]
CandlepinPoolManager revokeEntitlements

**Comment by wottop** [comment]
buildbot: retest this please

**Comment by mstead** [comment]
@wottop I'm not sure if it exists yet or not, but I'd like to see a spec test that concurrently binds to a pool and ensures the correct quantities result. Something like a pool with quantity 10 and 5 or 6 threads binding.

File: server/src/main/java/org/candlepin/model/PoolCurator.java [hunk 1]
```diff
@@ -996,4 +996,22 @@ public boolean exists(Pool pool) {
...
+
+    public void calculateExportedForOwnersPools(Owner owner) {
+        String stmt = "update Pool p set p.exported = coalesce(" +
+            "(select sum(ent.quantity) FROM Entitlement ent, Consumer cons, ConsumerType ctype " +
+            "where ent.pool.id = p.id and ent.consumer.id = cons.id and cons.type.id = ctype.id " +
+            "and ctype.manifest = 'Y'),0) where p.owner = :owner";
+        Query q = currentSession().createQuery(stmt);
+        q.setParameter("owner", owner);
+        q.executeUpdate();
+    }
```
Comment (mstead) [inline server/src/main/java/org/candlepin/model/PoolCurator.java]: Should get some coverage around these methods in PoolCuratorTest

**Comment by mstead** [comment]
retest this please

File: server/spec/entitlement_resource_spec.rb [hunk 2]
```diff
@@ -193,4 +193,36 @@
...
+    sleep 1
+    t2 = Thread.new{register_and_consume(pool, "candlepin", 7)}
+    sleep 1
+    t3 = Thread.new{register_and_consume(pool, "system", 6)}
+    sleep 1
+    t4 = Thread.new{register_and_consume(pool, "candlepin", 11)}
+    t1.join
+    t2.join
+    t3.join
+    t4.join
```
Comment (mstead) [inline server/spec/entitlement_resource_spec.rb]: Are the sleeps necessary here? Thread.join will make the main thread wait for all the binds to complete.

**Comment by mstead** [comment]
@wottop I'd also be curious as to how this would behave in a multi-threaded case where a pool's available quantity has been exhausted. Is there a possibility of a negative quantity?

**Comment by vritant** [comment]
remove default mark

**Comment by barnabycourt** [comment]
@wottop Looks like there is an error in the migration for mysql that could result in a null value for the quantity_consumed & exported

**Comment by barnabycourt** [comment]
@mstead have all your questions been answered?

**Comment by mstead** [comment]
> @wottop I'd also be curious as to how this would behave in a multi-threaded case where a pool's available quantity has been exhausted. Is there a possibility of a negative quantity?
> 
> Create a pool and make enough concurrent binds to exhaust the pools quantity. Might be tricky to get the test right.

@wottop I'd like to see this in there if we can as I think it is a very important test case.

**Comment by wottop** [comment]
the test 'should end at zero quantity consumed when all consumers are unregistered' was my attempt at that.

**Comment by mstead** [comment]
> the test 'should end at zero quantity consumed when all consumers are unregistered' was my attempt at that.

Yep. Good point. How about the case where we could possibly over consume?

**Comment by wottop** [comment]
'should not allow over consumption in pool' test

**Comment by mstead** [comment]
> 'should not allow over consumption in pool' test

Ugh... I suck. Thanks for pointing that out.

Looks good to me then.

**Comment by barnabycourt** [comment]
Performance test systems were having problems when this last test run was done. I ran them manually and they passed.

**Pull request merged**